README.md
# MMM-GrocyLists

This is a module for the [MagicMirror²](https://github.com/MichMich/MagicMirror/).

Integrate with a [Grocy](https://github.com/grocy/grocy) server to display information from various widgets.

## Using the module

To use this module, add the following configuration block to the modules array in the `config/config.js` file:
```js
var config = {
    modules: [
        {
            module: 'MMM-GrocyLists',
            header: "Grocy", // Customize depending on Widget.
            config: {
                // See below for configurable options
            }
        }
    ]
}
```

## Configuration options

| Option           | Description
|----------------- |-----------
| `grocyURL`       | *Required* URL of your Grocy instance. **Do not** include trailing `/`.<br><br>**Type:** `string`
| `grocyAPIKey`    | *Required* API Key for your Grocy instance. Can be obtained off your instance at `[Instance URL]/manageapikeys`.<br><br>**Type:** `string`
| `widgetType`     | *Optional* The type of Widget to render. <br><br>**Type:** `string` <br>Default `chores`
| `proxyCORS`      | *Optional* Whether to use [cors-anywhere.herokuapp.com] to proxy for CORS purposes. Use if having trouble with API calls.  <br><br>**Type:** `boolean` <br>Default `false`
| `updateInterval` | *Optional* How often to update the Module. <br><br>**Type:** `int`(milliseconds)<br>Default 300000 milliseconds (5 Minutes).
| `retryDelay`     | *Optional* How long to wait to retry if error on API call.<br><br>**Type:** `int`(milliseconds) <br>Default 5000 milliseconds (5 minute).
| `tableClass`     | *Optional* Class applied to the table. Controls sizing <br><br> `string`<br> Default `small`

## Widget Types
Currently only one widget type.

### Chores
Show a list of upcoming chores due, or those overdue.

### Thanks
* Built with [MagicMirror Module Template](https://github.com/roramirez/MagicMirror-Module-Template)


CHANGELOG.md
# MMM-GrocyLists Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).


## [1.0.0] - Chore Widget

- First public release. Chore Widget implemented. Formatting based on "Calendar" default extension.
- No filtering on Chore widget, will only show what the API returns.


Gruntfile.js
module.exports = function (grunt) {
  require('time-grunt')(grunt);
  grunt.initConfig({
    pkg:    grunt.file.readJSON('package.json'),
    eslint: {
      options: {
        configFile: '.eslintrc.json',
      },
      target: ['*.js'],
    },
    stylelint: {
      simple: {
        options: {
          configFile: '.stylelintrc',
        },
        src: ['*.css'],
      },
    },
    jsonlint: {
      main: {
        src:     ['package.json', 'translations/*.json'],
        options: {
          reporter: 'jshint',
        },
      },
    },
    markdownlint: {
      all: {
        options: {
          config: {
            default:                 true,
            'line-length':           false,
            'blanks-around-headers': false,
            'no-duplicate-header':   false,
            'no-inline-html':        false,
            MD010:                   false,
            MD001:                   false,
            MD031:                   false,
            MD040:                   false,
            MD002:                   false,
            MD029:                   false,
            MD041:                   false,
            MD032:                   false,
            MD036:                   false,
            MD037:                   false,
            MD009:                   false,
            MD018:                   false,
            MD012:                   false,
            MD026:                   false,
            MD038:                   false,
          },
        },
        src: ['README.md', 'CHANGELOG.md', 'LICENSE.txt'],
      },
    },
    yamllint: {
      all: ['.travis.yml'],
    },
  });
  grunt.loadNpmTasks('grunt-eslint');
  grunt.loadNpmTasks('grunt-stylelint');
  grunt.loadNpmTasks('grunt-jsonlint');
  grunt.loadNpmTasks('grunt-yamllint');
  grunt.loadNpmTasks('grunt-markdownlint');
  grunt.registerTask('default', ['eslint', 'stylelint', 'jsonlint', 'markdownlint', 'yamllint']);
};


node_helper.js
/* Magic Mirror
 * Node Helper: MMM-GrocyLists
 *
 * By Jordan Welch
 * MIT Licensed.
 */

const NodeHelper = require('node_helper');




module.exports = NodeHelper.create({

  // Override socketNotificationReceived method.

  /* socketNotificationReceived(notification, payload)
     * This method is called when a socket notification arrives.
     *
     * argument notification string - The identifier of the noitication.
     * argument payload mixed - The payload of the notification.
     */
  socketNotificationReceived(notification, payload) {
    if (notification === 'MMM-GrocyLists-NOTIFICATION_TEST') {
      console.log('Working notification system. Notification:', notification, 'payload: ', payload);
      // Send notification
      this.sendNotificationTest(this.anotherFunction()); // Is possible send objects :)
    }
  },

  // Example function send notification test
  sendNotificationTest(payload) {
    this.sendSocketNotification('MMM-GrocyLists-NOTIFICATION_TEST', payload);
  },

  // this you can create extra routes for your module
  extraRoutes() {
    const self = this;
    this.expressApp.get('/MMM-GrocyLists/extra_route', (req, res) => {
      // call another function
      values = self.anotherFunction();
      res.send(values);
    });
  },

  // Test another function
  anotherFunction() {
    return { date: new Date() };
  },
});


MMM-GrocyLists.js
/* eslint-disable no-restricted-syntax */
/* global Module */

/* Magic Mirror
 * Module: MMM-GrocyLists
 *
 * By Jordan Welch
 * MIT Licensed.
 */

Module.register('MMM-GrocyLists', {
  defaults: {
    grocyURL:       '', // The API of your Grocy instance.
    grocyAPIKey:    '', // The API Key of your Grocy Instance
    widgetType:     'chores',
    proxyCORS:      false,
    updateInterval: 300000, // 10 Minutes
    retryDelay:     5000,
    tableClass:     'small',
  },

  requiresVersion: '2.1.0', // Required version of MagicMirror

  start() {
    const self = this;
    const dataRequest = null;
    const dataNotification = null;

    // Flag for check if module is loaded
    this.loaded = false;

    // Schedule update timer.
    this.getData();
    setInterval(() => {
      self.updateDom();
    }, this.config.updateInterval);
  },
  /*
  * Format the URL for a given API Endpoint. Will pass trhough CORS proxy
  * if specified in settings.
  */
  APIEndpoint(endpoint) {
    if (this.config.proxyCORS) {
      return `https://cors-anywhere.herokuapp.com/${this.config.grocyURL}/api/${endpoint}`;
    }
    return `${this.config.grocyURL}/api/${endpoint}`;
  },
  /*
     * getData
     * function example return data and show it in the module wrapper
     * get a URL request
     *
     */
  getData() {
    const self = this;

    let retry = true;

    const dataRequest = new XMLHttpRequest();
    dataRequest.open('GET', self.APIEndpoint('chores'), true);
    dataRequest.setRequestHeader('GROCY-API-KEY', this.config.grocyAPIKey);
    dataRequest.onreadystatechange = function () {
      if (this.readyState === 4) {
        if (this.status === 200) {
          self.processData(JSON.parse(this.response));
        } else if (this.status === 401) {
          self.updateDom(self.config.animationSpeed);
          Log.error(self.name, this.status);
          retry = false;
        } else {
          Log.error(self.name, 'Could not load data.');
        }
        if (retry) {
          self.scheduleUpdate((self.loaded) ? -1 : self.config.retryDelay);
        }
      }
    };
    dataRequest.send();
  },

  /* scheduleUpdate()
     * Schedule next update.
     *
     * argument delay number - Milliseconds before next update.
     *  If empty, this.config.updateInterval is used.
     */
  scheduleUpdate(delay) {
    let nextLoad = this.config.updateInterval;
    if (typeof delay !== 'undefined' && delay >= 0) {
      nextLoad = delay;
    }
    nextLoad = nextLoad;
    const self = this;
    setTimeout(() => {
      self.getData();
    }, nextLoad);
  },

  createChoreList() {
    const chores = [];
    const NULLDATE = '2999-12-31 23:59:59';
    if (this.dataRequest) {
      // const today = moment().startOf('day');
      // const now = new Date();

      for (const c in this.dataRequest) {
        if (Object.prototype.hasOwnProperty.call(this.dataRequest, c)) {
          const drChore = this.dataRequest[c];

          if (drChore.next_estimated_execution_time !== NULLDATE) {
            const chore = {};

            chore.title = drChore.chore_name;
            chore.startDate = Date.parse(drChore.next_estimated_execution_time);

            chores.push(chore);
          }
        }
      }
    }
    return chores;
  },

  /* capFirst(string)
	 * Capitalize the first letter of a string
	 * Return capitalized string
	 */
  capFirst(string) {
    return string.charAt(0).toUpperCase() + string.slice(1);
  },

  getDom() {
    const self = this;

    const wrapper = document.createElement('table');
    // If this.dataRequest is not empty

    const chores = this.createChoreList();

    wrapper.className = this.config.tableClass;

    if (chores.length === 0) {
      wrapper.innerHTML = this.loaded ? this.translate('EMPTY') : this.translate('LOADING');
      wrapper.className = `${this.config.tableClass} dimmed`;
      return wrapper;
    }

    for (const c in chores) {
      if (Object.prototype.hasOwnProperty.call(chores, c)) {
        const chore = chores[c];
        const choreWrapper = document.createElement('tr');

        choreWrapper.className = 'normal';

        const titleWrapper = document.createElement('td');
        titleWrapper.innerHTML = chore.title;
        titleWrapper.className = 'title bright';

        const timeWrapper = document.createElement('td');
        choreWrapper.appendChild(titleWrapper);

        const now = new Date();
        // Define second, minute, hour, and day variables
        const oneSecond = 1000; // 1,000 milliseconds
        const oneMinute = oneSecond * 60;
        const oneHour = oneMinute * 60;
        const oneDay = oneHour * 24;

        if (chore.today) {
          timeWrapper.innerHTML = this.capFirst(this.translate('TODAY'));
        } else if (chore.startDate - now < oneDay && chore.startDate - now > 0) {
          timeWrapper.innerHTML = this.capFirst(this.translate('TOMORROW'));
        } else if (chore.startDate - now < 2 * oneDay && chore.startDate - now > 0) {
          if (this.translate('DAYAFTERTOMORROW') !== 'DAYAFTERTOMORROW') {
            timeWrapper.innerHTML = this.capFirst(this.translate('DAYAFTERTOMORROW'));
          } else {
            timeWrapper.innerHTML = this.capFirst(moment(chore.startDate, 'x').fromNow());
          }
        } else {
          timeWrapper.innerHTML = this.capFirst(moment(chore.startDate, 'x').from(moment().format('YYYYMMDD')));
        }
        choreWrapper.appendChild(timeWrapper);
        wrapper.appendChild(choreWrapper);
      }
    }
    return wrapper;
  },

  getScripts() {
    return [];
  },

  getStyles() {
    return [
      'MMM-GrocyLists.css',
    ];
  },

  // Load translations files
  getTranslations() {
    // FIXME: This can be load a one file javascript definition
    return {
      en: 'translations/en.json',
      es: 'translations/es.json',
    };
  },

  processData(data) {
    const self = this;
    this.dataRequest = data;
    if (this.loaded === false) { self.updateDom(self.config.animationSpeed); }
    this.loaded = true;

    // the data if load
    // send notification to helper
    this.sendSocketNotification('MMM-GrocyLists-NOTIFICATION_TEST', data);
  },

  // socketNotificationReceived from helper
  socketNotificationReceived(notification, payload) {
    if (notification === 'MMM-GrocyLists-NOTIFICATION_TEST') {
      // set dataNotification
      this.dataNotification = payload;
      this.updateDom();
    }
  },
});
